iroh-resolver/src/unixfs_builder.rs

```diff
@@ -197,74 +187,197 @@ impl File {
+            let name = match self.name {
+                Some(n) => n,
+                None => path
+                    .file_name()
+                    .and_then(|s| s.to_str())
+                    .unwrap_or_default()
+                    .to_string(),
+            };
+            return Ok(File {
+                content: Content::Path(path),
+                name,
+                chunker: Chunker::fixed_size(),
+                tree_builder: TreeBuilder::balanced_tree(),
+            });
+        }
+
+        if let Some(reader) = self.reader {
+            let name = self.name.ok_or_else(|| {
+                anyhow::anyhow!("must add a name when building a file from a reader or bytes")
+            })?;
+
+            return Ok(File {
+                content: Content::Reader(reader),
+                name,
+                chunker: Chunker::fixed_size(),
+                tree_builder: TreeBuilder::balanced_tree(),
+            });
+        }
+        anyhow::bail!("must have a path to the content or a reader for the content");
+    }
+}
+
+/// Entry is the kind of entry in a directory can be either a file or a
+/// folder (if recursive directories are allowed)
+#[derive(Debug, PartialEq)]
+enum Entry {
+    File(File),
+    Directory(Directory),
+}
+
+/// Construct a UnixFS directory.
+#[derive(Debug)]
+pub struct DirectoryBuilder {
+    name: Option<String>,
+    entries: Vec<Entry>,
+    typ: DirectoryType,
+    recursive: bool,
+}
+
+impl Default for DirectoryBuilder {
+    fn default() -> Self {
+        Self {
+            name: None,
+            entries: Default::default(),
+            typ: DirectoryType::Basic,
+            recursive: false,
+        }
+    }
+}
 
-        let name = self.name.ok_or_else(|| anyhow!("missing name"))?;
-        let reader = self.content.ok_or_else(|| anyhow!("missing content"))?;
+impl DirectoryBuilder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn as_hamt(&mut self) -> &mut Self {
+        self.typ = DirectoryType::Hamt;
+        self
+    }
+
+    pub fn allow_recursive(&mut self) -> &mut Self {
+        self.recursive = true;
+        self
+    }
+
+    pub fn with_name<N: Into<String>>(&mut self, name: N) -> &mut Self {
+        self.name = Some(name.into());
+        self
+    }
+
+    pub fn add_dir(&mut self, dir: Directory) -> Result<&mut Self> {
+        if !self.recursive {
+            anyhow::bail!("recursive directories not allowed");
+        }
+
+        if self.typ == DirectoryType::Basic && self.entries.len() >= DIRECTORY_LINK_LIMIT {
+            self.typ = DirectoryType::Hamt;
+        }
+        self.entries.push(Entry::Directory(dir));
+        Ok(self)
+    }
+
+    pub fn add_file(&mut self, file: File) -> &mut Self {
+        if self.typ == DirectoryType::Basic && self.entries.len() >= DIRECTORY_LINK_LIMIT {
+            self.typ = DirectoryType::Hamt
+        }
+        self.entries.push(Entry::File(file));
```

Comment:
maybe have a shared `fn push_entry` to have the HAMT size check in a single place